src/Tools/AI Test Toolkit/src/Logs/AITLogEntry.Codeunit.al

```diff
@@ -5,21 +5,163 @@
+        AITTestMethodLines: Page "AIT Test Method Lines Lookup";
+    begin
+        AITTestMethodLine.SetRange("Test Suite Code", TestSuiteCode);
+
+        AITTestMethodLines.SetTableView(AITTestMethodLine);
+        AITTestMethodLines.LookupMode(true);
+
+        if not (AITTestMethodLines.RunModal() = Action::LookupOK) then
+            exit;
+
+        AITTestMethodLines.GetRecord(AITTestMethodLine);
+
+        AITTestMethodLine.CalcFields("Codeunit Name");
+        LineNoFilter := StrSubstNo(LineNoFilterLbl, AITTestMethodLine."Codeunit ID", AITTestMethodLine."Codeunit Name", AITTestMethodLine."Input Dataset");
+        LineNo := AITTestMethodLine."Line No.";
+    end;
+
+    procedure UpdateTestInput(TestInput: Text; TestInputView: Enum "AIT Test Input - View"): Text
+    var
+        TestData: Codeunit "Test Input Json";
+    begin
+        InitTestData(TestInput, TestData);
+
+        case TestInputView of
+            TestInputView::"Full Input":
+                exit(TestInput);
+            TestInputView::Question:
+                exit(GetTestDataElement('question', TestData));
+            TestInputView::Context:
+                exit(GetTestDataElement('context', TestData));
+            TestInputView::"Test Setup":
+                exit(GetTestDataElement('test_setup', TestData));
+            TestInputView::"Ground Truth":
+                exit(GetTestDataElement('ground_truth', TestData));
+            TestInputView::"Expected Data":
+                exit(GetTestDataElement('expected_data', TestData));
+            else
+                exit('');
+        end;
+    end;
+
+    procedure UpdateTestOutput(TestOutput: Text; TestOutputView: Enum "AIT Test Output - View"): Text
+    var
+        TestData: Codeunit "Test Input Json";
+    begin
+        InitTestData(TestOutput, TestData);
+
+        case TestOutputView of
+            TestOutputView::"Full Output":
+                exit(TestOutput);
+            TestOutputView::Answer:
+                exit(GetTestDataElement('answer', TestData));
+            TestOutputView::Question:
+                exit(GetTestDataElement('question', TestData));
+            TestOutputView::Context:
+                exit(GetTestDataElement('context', TestData));
+            TestOutputView::"Ground Truth":
+                exit(GetTestDataElement('ground_truth', TestData));
+            else
+                exit('');
+        end;
+    end;
+
+    local procedure InitTestData(TestDataText: Text; var TestData: Codeunit "Test Input Json")
+    begin
+        if TestDataText = '' then
+            TestData.Initialize()
+        else
+            TestData.Initialize(TestDataText);
+    end;
+
+    local procedure GetTestDataElement(ElementName: Text; TestData: Codeunit "Test Input Json"): Text
+    var
+        ElementTestDataJson: Codeunit "Test Input Json";
+        ElementExists: Boolean;
+    begin
+        ElementTestDataJson := TestData.ElementExists('turns', ElementExists);
```

Comment:
This function is totally opposite once expectation. It saus "ElementExists" so i would think it returns a bool. 
but no, it returns data? and also takes a bool param.

Reply:
The bool param is passed by var and set to true/false depending on whether the element exist. Then the data is also returned if it exist. I didn't implement this functionality, simply using existing functionality in data-driven framework.